### PR TITLE
Implement commit search features for git

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

Implements the required features for the commit search plugin for git.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

**Contributor**:
- [x] PR is well described and the description can be used as a commit message on squash
- [ ] Related issues linked to PR if existing and labels set
- [x] New code is covered with unit tests
- [ ] New/updated ui components are tested inside the storybook (module ui-components only) 
- [x] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog`
- [ ] Feature has been tested with different permissions
- [ ] Documentation updated (only necessary for new features or changed behaviour)

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitHookChangesetProvider.java [hunk 1]
```diff
@@ -59,7 +53,8 @@ public GitHookChangesetProvider(GitChangesetConverterFactory converterFactory, R
   public synchronized HookChangesetResponse handleRequest(HookChangesetRequest request) {
     if (response == null) {
       GitHookChangesetCollector collector = new GitHookChangesetCollector(converterFactory, receivePack, receiveCommands);
-      response = new HookChangesetResponse(collector.collectChangesets());
+      collector.collectChangesets();
```
Comment (Pilopa) [inline scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitHookChangesetProvider.java:56]: I do not particularly like this "side-effect only" call, but I presume this is legacy code copied from somewhere ? The GitHookChangesetCollector looks like it was copied from somewhere.

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitHookChangesetCollector.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  Iterable<Changeset> getRemovedChangesets() {
+    return unmodifiableCollection(removedChangesets.values());
+  }
+
+  void handle(Repository repository, RevWalk walk, GitChangesetConverter converter, ReceiveCommand rc, String ref) {
+    try {
+      if (!(GitUtil.isBranch(ref) || GitUtil.isTag(ref))) {
+        LOG.debug("skip ref {}, because it is neither branch nor tag", ref);
+      } else if (rc.getType() == ReceiveCommand.Type.UPDATE_NONFASTFORWARD) {
+        LOG.debug("handle deleted ref {}", ref);
```
Comment (Pilopa) [inline scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitHookChangesetCollector.java:111]: I think this log should state something like "overwritten/replaced", or am I looking at this wrong ?

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitChangesetsCommand.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  GitChangesetsCommand(GitContext context, GitChangesetConverterFactory converterFactory) {
+    super(context);
+    this.converterFactory = converterFactory;
+  }
+
+  @Override
+  public Iterable<Changeset> getChangesets(ChangesetsCommandRequest request) {
+    try {
+      log.debug("computing changesets for repository {}", repository);
+      Repository repository = open();
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitChangesetsCommand.java:58]: The context and therefore the repository will be closed, when the service provider has finished. Closing the repository here might lead to errors, when other commands are called.

File: scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitChangesetsCommand.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+      }
+    } catch (IOException e) {
+      throw new InternalRepositoryException(context.getRepository(), "failed to get latest commit", e);
+    }
+  }
+
+  @Override
+  public Optional<Changeset> getLatestChangeset() {
+    try {
+      Repository repository = open();
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitChangesetsCommand.java:93]: Happens automatially, see above...

**Pull request merged**